File: backend/app/models/blog.py
"""
Blog Models Module

SQLAlchemy ORM models for blog article management and user interactions.

This module defines two core models:
- Article: Blog posts with AI-structured content
- SavedArticle: User bookmarks for articles

Key Features:
- AI-powered content structuring (structured_content JSON field)
- Multi-image support via JSON array
- SEO-friendly slug URLs
- Category-based organization
- Publication status control

Database Tables: articles, saved_articles
"""

from sqlalchemy import Column, Integer, String, Text, Boolean, ForeignKey, DateTime, JSON
from sqlalchemy.orm import relationship
from datetime import datetime
from app.core.database import Base

class Article(Base):
    """
    Blog article model with AI-powered content structuring.
    
    Supports rich content management with automatic section organization,
    multiple images, and SEO-optimized URLs.
    
    Attributes:
        id (int): Unique article identifier (primary key)
        title (str): Article headline (indexed for search)
        slug (str): URL-friendly identifier (unique, e.g., "tenerife-beaches")
        content (Text): Full article body (plain text or markdown)
        excerpt (str): Brief summary for listings (max 500 chars)
        category (str): Article classification (max 50 chars, indexed)
        image_url (str): DEPRECATED - Legacy single image URL
        image_slug (str): DEPRECATED - Legacy image identifier
        images (JSON): Array of image paths ["blog/img1.jpg", "blog/img2.jpg"]
        structured_content (JSON): AI-generated content sections with headers
        author_id (int): Article creator's user ID (nullable)
        is_published (bool): Publication status (default: False)
        created_at (DateTime): Article creation timestamp (UTC)
        
    Database Configuration:
        - Unique constraint on slug for SEO URLs
        - Indexes on id, title, slug, category for performance
        - JSON fields for flexible content storage
        
    Technical Notes:
        - structured_content format: {"sections": [{"title": "...", "content": "..."}]}
        - images field supports multiple image paths
        - Foreign key constraints removed for SQLite compatibility
        - Default unpublished status requires explicit publication
    """
    __tablename__ = "articles"
    id = Column(Integer, primary_key=True, index=True)
    title = Column(String, index=True)
    slug = Column(String, unique=True, index=True)
    content = Column(Text)
    excerpt = Column(String(500))  # Breve riassunto per liste
    category = Column(String(50), index=True)  # Categoria dell'articolo
    image_url = Column(String)  # URL completo immagine (opzionale) - deprecato, usare images
    image_slug = Column(String)  # slug per mappare l'immagine, es. 'teide' - deprecato, usare images
    images = Column(JSON, nullable=True)  # Array di path immagini, es. ['blog/teide-1.jpg', 'blog/teide-2.jpg']
    structured_content = Column(JSON, nullable=True)  # Struttura intelligente generata da AI
    author_id = Column(Integer, nullable=True)  # Author of the article (FK removed for compatibility)
    is_published = Column(Boolean, default=False)  # Published status
    created_at = Column(DateTime, default=datetime.utcnow)

class SavedArticle(Base):
    """
    User bookmark relationship for saved/favorited articles.
    
    Junction table implementing many-to-many relationship between
    users and articles for bookmark functionality.
    
    Attributes:
        id (int): Unique bookmark record identifier (primary key)
        user_id (int): User who saved the article
        article_id (int): Article that was saved
        
    Database Configuration:
        - Composite uniqueness enforced in application layer
        - Foreign key constraints removed for SQLite compatibility
        - Indexes recommended on user_id and article_id for queries
        
    Usage:
        - Query user's saved articles: filter by user_id
        - Count saves for article: count records by article_id
        - Check if user saved article: query both user_id and article_id
        
    Technical Notes:
        - No explicit timestamp (can be added if needed)
        - Deletion cascades handled in application layer
        - Supports quick bookmark/unbookmark operations
    """
    __tablename__ = "saved_articles"
    id = Column(Integer, primary_key=True, index=True)
    user_id = Column(Integer)  # FK removed for compatibility
    article_id = Column(Integer)  # FK removed for compatibility

File: backend/src/models/blog.ts
/**
 * Blog Models Module
 *
 * Article and SavedArticle models matching Python backend structure.
 *
 * Database Tables:
 * - articles: Blog posts with AI-structured content
 * - saved_articles: User bookmarks junction table
 */

import { DataTypes, Model, Optional } from "sequelize";
import { sequelize } from "../core/database";

/**
 * Article attributes interface
 * Matches Python Article model exactly
 */
interface ArticleAttributes {
  id: number;
  title: string;
  slug: string;
  content: string;
  excerpt: string | null;
  category: string | null;
  language: string;
  image_url: string | null;
  image_slug: string | null;
  images: string[] | null;
  structured_content: any | null;
  author_id: number | null;
  is_published: boolean;
  created_at: Date;
}

interface ArticleCreationAttributes
  extends Optional<
    ArticleAttributes,
    | "id"
    | "excerpt"
    | "category"
    | "image_url"
    | "image_slug"
    | "images"
    | "structured_content"
    | "author_id"
    | "is_published"
    | "created_at"
  > {}

/**
 * Article model class
 */
export class Article
  extends Model<ArticleAttributes, ArticleCreationAttributes>
  implements ArticleAttributes
{
  public id!: number;
  public title!: string;
  public slug!: string;
  public content!: string;
  public excerpt!: string | null;
  public category!: string | null;
  public language!: string;
  public image_url!: string | null;
  public image_slug!: string | null;
  public images!: string[] | null;
  public structured_content!: any | null;
  public author_id!: number | null;
  public is_published!: boolean;
  public created_at!: Date;
}

/**
 * Initialize Article model
 */
Article.init(
  {
    id: {
      type: DataTypes.INTEGER,
      autoIncrement: true,
      primaryKey: true,
    },
    title: {
      type: DataTypes.STRING,
      allowNull: false,
    },
    slug: {
      type: DataTypes.STRING,
      allowNull: false,
      unique: true,
    },
    content: {
      type: DataTypes.TEXT,
      allowNull: false,
    },
    excerpt: {
      type: DataTypes.STRING(500),
      allowNull: true,
    },
    category: {
      type: DataTypes.STRING(50),
      allowNull: true,
    },
    language: {
      type: DataTypes.STRING(10),
      allowNull: false,
      defaultValue: "es",
    },
    image_url: {
      type: DataTypes.STRING,
      allowNull: true,
    },
    image_slug: {
      type: DataTypes.STRING,
      allowNull: true,
    },
    images: {
      type: DataTypes.JSON,
      allowNull: true,
    },
    structured_content: {
      type: DataTypes.JSON,
      allowNull: true,
    },
    author_id: {
      type: DataTypes.INTEGER,
      allowNull: true,
    },
    is_published: {
      type: DataTypes.BOOLEAN,
      defaultValue: false,
    },
    created_at: {
      type: DataTypes.DATE,
      defaultValue: DataTypes.NOW,
    },
  },
  {
    sequelize,
    tableName: "articles",
    timestamps: false,
  }
);

/**
 * SavedArticle model for user bookmarks
 */
interface SavedArticleAttributes {
  id: number;
  user_id: number;
  article_id: number;
}

interface SavedArticleCreationAttributes
  extends Optional<SavedArticleAttributes, "id"> {}

export class SavedArticle
  extends Model<SavedArticleAttributes, SavedArticleCreationAttributes>
  implements SavedArticleAttributes
{
  public id!: number;
  public user_id!: number;
  public article_id!: number;
}

SavedArticle.init(
  {
    id: {
      type: DataTypes.INTEGER,
      autoIncrement: true,
      primaryKey: true,
    },
    user_id: {
      type: DataTypes.INTEGER,
      allowNull: false,
    },
    article_id: {
      type: DataTypes.INTEGER,
      allowNull: false,
    },
  },
  {
    sequelize,
    tableName: "saved_articles",
    timestamps: false,
  }
);
